File: 30Days6Companies/Microsoft/150. Evaluate Reverse Polish Notation/sol.cpp
class Solution {
public:
    bool is_oper(string p){
        if(p=="+"||p=="-"||p=="*"||p=="/")
            return true;
        return false;
    }
    int evalRPN(vector<string>& tokens) {
        stack<int> stk;
        char ch;
        int i=0;
        int result,op1,op2;
        for (int i = 0; i < tokens.size(); i++)
        {
            string ch=tokens[i];
            if(is_oper(ch))
            {
                op1=stk.top();
                stk.pop();
                op2=stk.top();
                stk.pop();
                
                if (ch=="+")
                {
                    result=op1+op2;
                    stk.push(result);
                }
                
                else if(ch=="-")
                {
                    result=op2-op1;
                    stk.push(result);
                }
                else if(ch=="*")
                {
                    result=op1*op2;
                    stk.push(result);
                }
                else if(ch=="/")
                {
                    result=op2/op1;
                    stk.push(result);
                }
            }
            else
            {
                stk.push(stoi(ch));
            }
        }
           result=stk.top();
            return result; 
        }
        
    
};

File: 30Days6Companies/Microsoft/150. Evaluate Reverse Polish Notation/sol.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stk=[]
        for token in tokens:
            if(token in "*/+-"):
                op1=stk[-1]
                stk.pop()
                op2=stk[-1]
                stk.pop()
                if(token=="+"):
                    result=op1+op2
                    stk.append(result)
                elif(token=="-"):
                    result=op2-op1
                    stk.append(result)
                elif(token=="*"):
                    result=op2*op1
                    stk.append(result)
                elif(token=="/"):
                    result=int(op2/op1)
                    stk.append(result)
            else:
                stk.append(int(token))

        
        return stk[-1]
